lib/rubocop/cop/performance/use_zip_to_wrap_array_contents.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Performance
+      # Checks for `.map { |id| [id] }` and suggests replacing it with `.zip`.
+      #
+      # @example
+      #   # bad
+      #   [1, 2, 3].map { |id| [id] }
+      #
+      #   # good
+      #   [1, 2, 3].zip
+      #
+      # @example
+      #   # good (no offense)
+      #   [1, 2, 3].map { |id| id }
+      #
+      # @example
+      #   # good (no offense)
+      #   [1, 2, 3].map { |id| [id, id] }
+      class UseZipToWrapArrayContents < Base
+        include RangeHelp
+        extend AutoCorrector
+
+        MSG = 'Use `.zip` instead of `.map { |id| [id] }`.'
+        RESTRICT_ON_SEND = %i[map].freeze
+
+        # Matches regular block form `.map { |e| [e] }`
+        # @!method map_with_array?(node)
+        def_node_matcher :map_with_array?, <<~PATTERN
+          (block
+            (send _ :map)
+            (args (arg _id))
+            (array (lvar _id)))
+        PATTERN
+
+        # Matches numblock form `.map { [_1] }`
+        # @!method map_with_array_numblock?(node)
+        def_node_matcher :map_with_array_numblock?, <<~PATTERN
+          (numblock
+            (send _ :map)
+            1
+            (array (lvar _))
+          )
+        PATTERN
+
+        def on_send(node)
+          return unless (parent = node.parent)
```

Comment:
I'm not opposed to this implementation as you have something that is working. If my memory serves correctly, it tends to be better to start with the parent as the node and extract the children out. I think this would require splitting `on_send` into `on_block` and `on_numblock`.

Reply:
That's actually what we started with in this PR. @Earlopain asked to refactor to the `on_send` approach. 

Thank you for this comment though, it really helps contextualize some of the trial and error that went into getting it to work this way. 

I do like the current approach a lot now, it feels very clean to express it in Ruby with guard clauses.

```
module RuboCop
  module Cop
    module Performance
      # Checks for `.map { |id| [id] }` and suggests replacing it with `.zip`.
      
      class UseZipToWrapArrayContents < Base
        extend RuboCop::Cop::AutoCorrector

        MSG = 'Use `.zip` instead of `.map { |id| [id] }`.'

        # @!method map_with_array?(node)
        def_node_matcher :map_with_array?, <<~PATTERN
          (block
            (send _ :map)
            (args (arg _id))
            (array (lvar _id))
          )
        PATTERN

        # Matcher for numblock (shorthand block with _1)
        # @!method map_with_array_numblock?(node)
        def_node_matcher :map_with_array_numblock?, <<~PATTERN
          (numblock
            (send _ :map)
            1
            (array (lvar _))
          )
        PATTERN

        def on_block(node)
          return unless node.method?(:map)
          return unless map_with_array?(node)

          add_offense(node) do |corrector|
            autocorrect(node).call(corrector)
          end
        end

        def on_numblock(node)
          return unless node.method?(:map)
          return unless map_with_array_numblock?(node)

          add_offense(node) do |corrector|
            autocorrect(node).call(corrector)
          end
        end

        def autocorrect(node)
          -> (corrector) do
            corrector.replace(node, correct_replacement(node))
          end
        end

        private

        def correct_replacement(node)
          map_call = node.children.first
          receiver = map_call.receiver.source
          "#{receiver}.zip"
        end
      end
    end
  end
end

```